### PR TITLE
UNR-537: Unreal GDK as a plugin

### PR DESCRIPTION
**Pull request by danielimprobable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
This PR restructures the GDK so that it can be built as a uplugin which simplifies the installation process.
#### Tests
This PR has been tested agains a branch of [testsuite](https://github.com/spatialos/UnrealGDKTestSuite/pull/103)
#### Documentation
Additional PR with the docs will be created once merged.
#### Primary reviewers
@improbable-valentyn @joshuahuburn

### CODE REVIEW

File: SpatialGDK/SpatialGDK.uplugin [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	"VersionName": "1.0",
+	"FriendlyName": "SpatialGDK",
+	"Description": "SpatialOS Unreal GDK",
+	"Category": "SpatialOS Unreal GDK",
+	"CreatedBy": "Improbable Worlds, Ltd.",
+	"CreatedByURL": "https://improbable.io",
+	"DocsURL": "",
+	"MarketplaceURL": "",
+	"SupportURL": "https://forums.improbable.io/c/sup",
+    "EnabledByDefault" : true,
```
Comment (m-samiec) [inline SpatialGDK/SpatialGDK.uplugin]: Formatting in this file.

File: SpatialGDKEditorToolbar/SpatialGDKEditorToolbar.uplugin [hunk 2]
```diff
@@ -17,7 +17,14 @@
 		{
```
Comment (m-samiec) [inline SpatialGDKEditorToolbar/SpatialGDKEditorToolbar.uplugin]: The support URL above is invalid. https://forums.improbable.io/ ?

File: SpatialGDK/SpatialGDK.uplugin [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	"CanContainContent": true,
+	"IsBetaVersion": false,
```
Comment (joshuahuburn) [inline SpatialGDK/SpatialGDK.uplugin]: Are we sure we're not a beta version :D

File: SpatialGDK/SpatialGDK.uplugin [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{
+	"FileVersion": 3,
+	"Version": 1,
+	"VersionName": "1.0",
+	"FriendlyName": "SpatialGDK",
+	"Description": "SpatialOS Unreal GDK",
+	"Category": "SpatialOS Unreal GDK",
+	"CreatedBy": "Improbable Worlds, Ltd.",
+	"CreatedByURL": "https://improbable.io",
+	"DocsURL": "",
```
Comment (joshuahuburn) [inline SpatialGDK/SpatialGDK.uplugin]: Maybe a link to our GitHub docs here?

File: ci/build_linux.bat [hunk 5]
```diff
@@ -0,0 +1,15 @@
+@echo off
+
+pushd "%~dp0../SpatialGDK"
+
+call Setup.bat
+
+popd
```
Comment (m-samiec) [inline ci/build_linux.bat]: Redundant pop and push?

File: SpatialGDKEditorToolbar/Source/Private/InteropCodeGenerator/SpatialGDKEditorInteropCodeGenerator.cpp [hunk 6]
```diff
@@ -234,7 +234,7 @@ bool SpatialGDKGenerateInteropCode()
 	const FString SchemaIntermediatePath = GenerateIntermediateDirectory();
 	GenerateInteropFromClasses(InteropGeneratedClasses, SchemaIntermediatePath);
 
-	const FString DiffCopyPath = FPaths::ConvertRelativePathToFull(FPaths::Combine(*FPaths::GetPath(FPaths::GetProjectFilePath()), TEXT("Scripts/DiffCopy.bat")));
+	const FString DiffCopyPath = FPaths::ConvertRelativePathToFull(FPaths::Combine(*FPaths::GetPath(FPaths::GetProjectFilePath()), TEXT("Plugins/UnrealGDK/SpatialGDK/Build/Scripts/DiffCopy.bat")));
```
Comment (joshuahuburn) [inline SpatialGDKEditorToolbar/Source/Private/InteropCodeGenerator/SpatialGDKEditorInteropCodeGenerator.cpp]: Can remove diff-copy now I believe.

File: SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar.Build.cs [hunk 7]
```diff
@@ -11,30 +11,26 @@ public SpatialGDKEditorToolbar(ReadOnlyTargetRules Target) : base(Target)
 
         PublicIncludePaths.AddRange(
             new string[] {
-                "SpatialGDKEditorToolbar/Public",
-                "SpatialGDK/Public"
```
Comment (joshuahuburn) [inline SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar.Build.cs]: How come we no longer require the SpatialGDK public includes?

File: .gitignore [hunk 8]
```diff
@@ -58,31 +58,30 @@ Build/*/**
...
 .spatialos/
-Plugins/SpatialOS
 logs/
 
 Scripts/spatialos.*.build.json
 
-!Source/Programs/Improbable.Unreal.CodeGeneration/UnrealCodeGeneration.sln
-!Source/Programs/Improbable.Unreal.Scripts/Improbable.Unreal.Scripts.sln
+!SpatialGDK/Build/Programs/Improbable.Unreal.CodeGeneration/UnrealCodeGeneration.sln
+!SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Improbable.Unreal.Scripts.sln
```
Comment (m-samiec) [inline .gitignore]: Does this only contain diffcopy now?

File: SpatialGDK/Setup.bat [hunk 9]
```diff
@@ -7,7 +7,7 @@ pushd "%~dp0"
 call :MarkStartOfBlock "%~0"
```
Comment (joshuahuburn) [inline SpatialGDK/Setup.bat]: Maybe this should be in the root directory and not in SpatialGDK. I think it's a little strange having it in a subfolder when there is SpatialGDK and SpatialGDKEditorToolbar.

File: .gitignore [hunk 10]
```diff
@@ -58,31 +58,30 @@ Build/*/**
...
 
 .spatialos/
-Plugins/SpatialOS
 logs/
 
 Scripts/spatialos.*.build.json
 
-!Source/Programs/Improbable.Unreal.CodeGeneration/UnrealCodeGeneration.sln
-!Source/Programs/Improbable.Unreal.Scripts/Improbable.Unreal.Scripts.sln
+!SpatialGDK/Build/Programs/Improbable.Unreal.CodeGeneration/UnrealCodeGeneration.sln
```
Comment (m-samiec) [inline .gitignore]: This is no longer required.

File: .gitignore (hunk 8 above)
Comment (m-samiec) [reply .gitignore]: Ah, build, linter and diffcopy.

File: SpatialGDK/SpatialGDK.uplugin (hunk 4 above)
Comment (danielimprobable) [reply SpatialGDK/SpatialGDK.uplugin]: Shall I link to the spatial docs?

File: SpatialGDK/Setup.bat (hunk 9 above)
Comment (joshuahuburn) [reply SpatialGDK/Setup.bat]: I see your point but to me it feels strange to have to go to a sub-folder to actually 'install' the plugin if you know what I mean.

**Comment by danielimprobable** [comment]
I've now successfully deployed this PR now

File: SpatialGDK/Documentation/setup-and-installing.md [hunk 11]
```diff
@@ -84,40 +84,38 @@ If the environment variable is registered correctly, this returns the path you u
 
 Follow the steps below to:
 
-* Clone the Unreal GDK and Starter Project repositories.
-* Build the Unreal GDK module dependencies which the Starter Project needs so it can work with the GDK, and add the Unreal GDK to the Starter Project.
+* Clone the Starter Project repository.
+* Clone the Unreal GDK into the `Plugins` folder of your GDK. (If you don't have any other plugins you will need to create the plugins folder).
```
Comment (m-samiec) [inline SpatialGDK/Documentation/setup-and-installing.md]: `into the Plugins folder of your game`?

File: SpatialGDK/Documentation/setup-and-installing.md [hunk 12]
```diff
@@ -84,40 +84,38 @@ If the environment variable is registered correctly, this returns the path you u
...
-    1. Drag the Unreal GDK folder onto the batch script **CreateGDKSymlinks.bat**.</br>This brings up a terminal window, and the output should be something like `Successfully created symlinks to “C:\Users\name\Documents\unreal-gdk”`.</br>For more information on helper scripts, see [Helper scripts](https://github.com/spatialos/UnrealGDKStarterProject#helper-scripts) in the Starter Project readme.
-1. Within the root directory of the Starter Project repository, go to **Game > Scripts** and double-click **`Codegen.bat`**. </br>This initializes the project. It should succeed quickly and silently.
-1. Set the Starter Project to work with the Unreal Engine fork you cloned and installed earlier. To do this:
+1. Open **File Explorer**, navigate to the root directory of the Unreal GDK repository, and double-click **`Setup.bat`**. You may be prompted to sign into your SpatialOS account if you have not already.
+2. Set the Starter Project to work with the Unreal Engine fork you cloned and installed earlier. To do this:
```
Comment (m-samiec) [inline SpatialGDK/Documentation/setup-and-installing.md]: I believe these should all be `1.`

File: SpatialGDK/Documentation/setup-and-installing.md [hunk 13]
```diff
@@ -84,40 +84,38 @@ If the environment variable is registered correctly, this returns the path you u
...
-1. Close the Unreal Editor and build the project again in Visual Studio.
-1. In File Explorer, navigate to the root directory of the Unreal GDK Starter Project repository, then to **`\Game\Scripts`**, and run **`BuildWorkerConfig.bat`**.
+    2. Right-click **StarterProject.uproject** and select **Switch Unreal Engine version**.
+    3. Select the path to the Unreal Engine fork you cloned earlier.
+3. In the same directory, open **StarterProject.sln** with Visual Studio.
+4. In the Solution Explorer window, right-click on **StarterProject** and select **Build**.
+5. Open **StarterProject.uproject** in the Unreal Editor and click [**Codegen**]({{urlRoot}}/content/interop) to generate [type bindings]({{urlRoot}}/content/glossary#type-bindings).
+6. Close the Unreal Editor and build the project again in Visual Studio.
+7.  In File Explorer, navigate to the root directory of the Unreal GDK Starter Project repository, then to **`\Game\Plugins\UnrealGDK\SpatialGDK\Build\Scripts`**, and run **`BuildWorkerConfig.bat`**.
```
Comment (m-samiec) [inline SpatialGDK/Documentation/setup-and-installing.md]: Tab/space mixup after `7.`

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Improbable.Unreal.Scripts.sln [hunk 14]
```diff
@@ -37,4 +31,7 @@ Global
 	GlobalSection(SolutionProperties) = preSolution
 		HideSolutionNode = FALSE
 	EndGlobalSection
+	GlobalSection(ExtensibilityGlobals) = postSolution
```
Comment (m-samiec) [inline SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Improbable.Unreal.Scripts.sln]: What is this for?

File: SpatialGDK/Documentation/setup-and-installing.md [hunk 15]
```diff
@@ -137,12 +135,12 @@ To run a cloud deployment, you need to prepare your server-worker and client-wor
...
 1. In a terminal window, navigate to the root directory of the Unreal GDK Starter Project repository.
-1. Build a server-worker assembly: `Game\Scripts\BuildWorker.bat StarterProjectServer Linux Development StarterProject.uproject`
-1. Build a client-worker assembly: `Game\Scripts\BuildWorker.bat StarterProject Win64 Development StarterProject.uproject`
-1. Navigate to `StarterProject\spatial`.
-1. Upload the assemblies to the cloud, specifying an assembly name (this covers both assemblies): `spatial cloud upload <assembly_name>`
-1. Launch a deployment, specifying a deployment name: `spatial cloud launch <assembly_name> one_worker_test.json <deployment_name> --snapshot=snapshots\default.snapshot`
-1. Follow the steps [here](https://docs.improbable.io/reference/latest/shared/get-started/tour#start-a-game-client) (SpatialOS documentation) to launch the game.
+2. Build a server-worker assembly: `Game\Plugins\UnrealGDK\SpatialGDK\Build\Scripts\BuildWorker.bat StarterProjectServer Linux Development StarterProject.uproject`
```
Comment (m-samiec) [inline SpatialGDK/Documentation/setup-and-installing.md]: These should all be `1.`

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Improbable.Unreal.Scripts.sln (hunk 14 above)
Comment (danielimprobable) [reply SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Improbable.Unreal.Scripts.sln]: Not sure added by vs2017 https://stackoverflow.com/questions/45861679/what-is-the-new-globalsection-in-a-vs2017-15-3-solution-file will see if I can remove it

File: SpatialGDK/SpatialGDK.uplugin [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+	"Installed": true,
+	"Modules": [
+		{
+			"Name": "SpatialGDK",
+			"Type": "Runtime",
+			"LoadingPhase": "Default",
+			"WhitelistPlatforms": [ "Win64", "Linux", "XboxOne", "PS4" ]
+		},
+	],
+	"Plugins": [
```
Comment (m-samiec) [inline SpatialGDK/SpatialGDK.uplugin]: Shouldn't it be up to the user to add these additional plugins?

Comment (danielimprobable) [reply SpatialGDK/SpatialGDK.uplugin]: These are dependencies for this plugin and has to be here

File: SpatialGDK/Documentation/setup-and-installing.md [hunk 17]
```diff
@@ -93,8 +93,8 @@ Follow the steps below to:
 1. Clone the [Unreal GDK Starter Project](https://github.com/spatialos/UnrealGDKStarterProject/) repository by running either:
     * (HTTPS) `git clone https://github.com/spatialos/UnrealGDKStarterProject.git`
     * (SSH) `git clone git@github.com:spatialos/UnrealGDKStarterProject.git`
-2. Navigate into `<StarterProjectRepoRoot>/Game/Plugins`. You may need to create the Plugins folder if you don't have any plugins installed yet.
-3.  In a Git Bash terminal window, clone the [Unreal GDK](https://github.com/spatialos/UnrealGDK) repository by running either:
+1. Navigate into `<StarterProjectRepoRoot>/Game/Plugins`. You may need to create the Plugins folder if you don't have any plugins installed yet.
```
Comment (joshuahuburn) [inline SpatialGDK/Documentation/setup-and-installing.md]: The top line of this `.md` seems to be broken. Can you do a drive-by fix?
